16_Trade-offs.Rmd

```diff
@@ -1,372 +1,394 @@
-speak <- new_generic("speak", "x")
-
-method(speak, dog) <- function(x) {
-  "Woof"
-}
-  
-speak(lola)
-
-#> [1] "Woof"
-```
-
-If we have another class, we can implement the generic for that too:
-
-```{r, eval = FALSE}
-cat <- new_class("cat", properties = list(
-  name = class_character,
-  age = class_double
-))
-method(speak, cat) <- function(x) {
-  "Meow"
-}
-
-fluffy <- cat(name = "Fluffy", age = 5)
-speak(fluffy)
-
-#> [1] "Meow"
-```
-
-Helpful messages:
-
-```{r, eval = FALSE}
-speak
-
-#> <S7_generic> speak(x, ...) with 2 methods:
-#> 1: method(speak, cat)
-#> 2: method(speak, dog)
-```
-
-
-"most usage of S7 with S3 will just work"
-
-```{r, eval = FALSE}
-method(print, cat) <- function(...) {
-  print("I am a cat.")
-}
-
-print(fluffy)
-#> "I am a cat"
-
-```
-
-*For validators, inheritance, dynamic properties and more,  see the [vignette!](https://rconsortium.github.io/OOP-WG/articles/S7.html)*
-
-https://rconsortium.github.io/OOP-WG/articles/S7.html
-
-## So... switch to S7 ? {-}
-
-$$
-\huge
-\textbf{Soon}^{tm}
-$$
-
-* Not yet... still in development! 
-
-* But consider trying it out:
-
-   * To stay ahead of the curve... S7 will be integrated into base R someday!
-   
-   * To contribute feedback to the S7 team!
-
-   * To get "almost all" of the benefits of S4 without the complexity !  
-   
-* In particular, if you have a new project that might require the complexity of S4, consider S7 instead!
-
-## Meeting Videos
-
-### Cohort 1
-
-`r knitr::include_url("https://www.youtube.com/embed/W1uc8HbyZvI")`
-
-### Cohort 2
-
-`r knitr::include_url("https://www.youtube.com/embed/bzo37PHCM1I")`
-
-### Cohort 3
-
-`r knitr::include_url("https://www.youtube.com/embed/_byYFTQHp1Y")`
-
-### Cohort 4
-
-`r knitr::include_url("https://www.youtube.com/embed/vdKDPBcOc6Y")`
-
-### Cohort 5
-
-`r knitr::include_url("https://www.youtube.com/embed/3EvqtVYTFVM")`
-
-### Cohort 6
-
-`r knitr::include_url("https://www.youtube.com/embed/vEButxFIvLw")`
-
-<details>
-<summary> Meeting chat log </summary>
-
-```
-00:11:36	Oluwafemi Oyedele:	I have not built anything with them!!!
-00:16:31	Arthur Shaw:	https://cran.r-project.org/web/packages/sp/index.html
-00:19:05	Arthur Shaw:	Apparently Hadley asked the same question we're asking several years ago: https://stackoverflow.com/questions/5437238/which-packages-make-good-use-of-s4-objects
-00:19:16	Trevin:	HA
-00:23:54	Trevin:	Your audio is breaking up Federica
-01:06:58	Federica Gazzelloni:	https://mastering-shiny.org/reactive-motivation.html?q=R6#event-driven
-01:07:37	Federica Gazzelloni:	https://engineering-shiny.org/common-app-caveats.html?q=R6#using-r6-as-data-storage
-01:10:52	Oluwafemi Oyedele:	Thank you !!!
-```
-</details>
-
-### Cohort 7
-
-`r knitr::include_url("https://www.youtube.com/embed/2vxnzqWp-OU")`
+# Trade-offs
+
+**Learning objectives:**
+
+- Understand the Trade-offs between S3, R6 and S4
+
+- Brief intro to S7 (the object system formerly known as R7)
+
+
+## Introduction
```

Comment:
```suggestion
## Introduction {-}
```
We don't want numbers in any of the sections, since slides won't necessarily match the book (and shouldn't in a lot of cases). This will become moot ~soon when we upgrade to Quarto, but for now I want to be explicit. I'll probably merge these changes myself if there aren't other things that need fixing.